File: Chapter_1/milk/milk.java

/*
ID: 100021881
LANG: JAVA
PROG: milk
*/

import java.io.*;
import java.util.*;

public class milk {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new FileReader("milk.in"));
        PrintWriter out = new PrintWriter(new BufferedWriter(new FileWriter("milk.out")));
        StringTokenizer l = new StringTokenizer(br.readLine());
        int total = Integer.parseInt(l.nextToken());
        int numFarmers = Integer.parseInt(l.nextToken());
        List<Integer> prices = new ArrayList<Integer>();
        List<Integer> amounts = new ArrayList<Integer>();
        
        for (int i = 0; i < numFarmers; i++) {
            StringTokenizer li = new StringTokenizer(br.readLine());
            int price = Integer.parseInt(li.nextToken());
            int amount = Integer.parseInt(li.nextToken());
            int j = 0;
            for (; j < prices.size() && prices.get(j) < price; j++) {}
            prices.add(j, price);
            amounts.add(j, amount);
        }
        int price = 0;
        int i = 0;
        for (i = 0; i < prices.size() && total > 0; i++) {
            total-=amounts.get(i);
            price+=prices.get(i)*amounts.get(i);
        }
        if (total!=0) {
            price+=total*prices.get(i-1);
            total-=total;
        }
        out.println(price);
        out.close();
        br.close();
    }
}

File: Chapter_1/milk/milk.cpp
/*
ID: 100021881
LANG: C++
PROG: milk
*/
#include <iostream>
#include <fstream>
#include <string>
#include <algorithm>
#include <vector>

using namespace std;

int main() {
    ifstream fin("milk.in");
    ofstream fout("milk.out");
    int total, numFarmers;
    fin >> total >> numFarmers;
    vector<pair<int, int>> farmers;
    for (int i = 0; i < numFarmers; i++) {
        int price, amount;
        fin >> price >> amount;
        farmers.push_back(make_pair(price, amount));
    }
    sort(farmers.begin(), farmers.end());
    int price = 0, i;
    for (i = 0; i < farmers.size() && total > 0; i++) {
        total-=farmers[i].second;
        price+=farmers[i].first*farmers[i].second;
    }
    if (total!=0) {
        price+=total*farmers[i-1].first;
    }
    fout << price << endl;
    exit(0);
}